src/Temporalio/Worker/WorkflowReplayer.cs

```diff
@@ -0,0 +1,270 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Temporalio.Bridge.Api.WorkflowActivation;
+using Temporalio.Exceptions;
+using Temporalio.Runtime;
+using Temporalio.Worker.Interceptors;
+
+#if NETCOREAPP3_0_OR_GREATER
+using System.Runtime.CompilerServices;
+#endif
+
+namespace Temporalio.Worker
+{
+    /// <summary>
+    /// Replayer to replay workflows from existing history.
+    /// </summary>
+    public class WorkflowReplayer
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WorkflowReplayer"/> class. Note, some
+        /// options validation is deferred until replay is attempted.
+        /// </summary>
+        /// <param name="options">Replayer options.</param>
+        public WorkflowReplayer(WorkflowReplayerOptions options)
+        {
+            if (options.Workflows.Count + options.AdditionalWorkflowDefinitions.Count == 0)
+            {
+                throw new ArgumentException("Must have at least one workflow");
+            }
+            Options = options;
+        }
+
+        /// <summary>
+        /// Gets the options this replayer was created with.
+        /// </summary>
+        public WorkflowReplayerOptions Options { get; private init; }
+
+        /// <summary>
+        /// Replay a single workflow from the given history.
+        /// </summary>
+        /// <param name="history">History to replay.</param>
+        /// <param name="throwOnReplayFailure">If true, the default, this will throw
+        /// <see cref="InvalidWorkflowOperationException" /> on a workflow task failure
+        /// (e.g. non-determinism). This has no effect on workflow failures which are not reported
+        /// as part of replay.</param>
+        /// <returns>Result of the replay run.</returns>
+        public async Task<WorkflowReplayResult> ReplayWorkflowAsync(
+            WorkflowHistory history, bool throwOnReplayFailure = true) =>
+            (await ReplayWorkflowsAsync(new[] { history }, throwOnReplayFailure).ConfigureAwait(false)).Single();
+
+        /// <summary>
+        /// Replay multiple workflows from the given histories.
```

Comment:
Might want to direct users to the async API if they're on a new enough version of .NET.